README
NAME
    WordNet-SenseRelate-WordToSet version 0.03

OVERVIEW
    This module takes as input a single target word, and a set of one or
    more other words. It finds the sense of that target word that is most
    related to those words in the set. For example, if the target word is
    "bank", and the words in the set are "money cash loan stock", we might
    expect that the most related sense of "bank" is that pertaining to
    financial instituations.

    This is potentially useful in determining the predominant sense of a
    word in a particular domain. For example, if the target word is "game",
    and the words in the set are from the domain of board games (e.g.,
    "monopoly chess checkers", then the sense of "game" that we'd expect to
    be most similar or related would be that of games you play rather than
    the game you hunt. For example, here's some output for this case:

     wordtoset.pl game monopoly checkers chess --type WordNet::Similarity::wup

     game#n#1 : 2.52631578947368 : a contest with rules to determine a winner; "you need four people to play this game"  

     game#n#10 : 2.17777777777778 : your occupation or line of work; "he's in the plumbing game"; "she's in show biz"  

     game#n#3 : 2.17777777777778 : an amusement or pastime; "they played word games"; "he thought of his painting as a game that filled his empty time"; "his life was all fun and games"  

     ------------------------

     wordtoset.pl game turkey boar deer --type WordNet::Similarity::wup

     game#n#4 : 1.98 : animal hunted for food or sport  

     game#n#7 : 1.27777777777778 : the flesh of wild animals that is used for food  

     game#n#9 : 1.24542124542125 : the game equipment needed in order to play a particular game; "the child received several games for his birthday"  

    Note that wordtoset.pl will output all of the senses, but we've only
    shown the top three here in the interests of brevity. We can see that
    according to the Wu-Palmer measure (wup), the sense of game most similar
    to the given sense is as we've described above.

    WordToSet might also be useful in detecting sentiment orientation. For
    example, suppose the target word is "war". You could compare that to two
    different sets such as : "peace love happiness" and "hate death fear".
    While the predominant sense of "war" might not change, if it has a
    substantially higher score relative to one of the sets then it could be
    concluded that war is more associated with that set than the other.

    This module uses WordNet and measures of semantic relatedness and
    similarity from WordNet::Similarity to arrive at its output.

SYNOPSIS
       # from the command line
   
       wordtoset.pl star nebula cosmos orion --type WordNet::Similarity::lin

       wordtoset.pl star movie hollywood director --type WordNet::Similarity::vector 

       # from within a program

       use WordNet::SenseRelate::WordToSet;
       use WordNet::QueryData;

       my $qd = WordNet::QueryData->new;

       my %options = (wordnet => $qd,
                      measure => 'WordNet::Similarity::lesk');

       my $wsd = WordNet::SenseRelate::WordToSet->new (%options);

       my $result = $wsd->disambiguate (target => 'java',
                          context => ['programming_language', 'applet']);

       foreach my $key (keys %$result) {
           print $key, ' : ', $result->{$key}, "\n";
       }

CONTENTS
    When the distribution is unpacked, several subdirectories are created:

    /lib
        This directory contains the Perl modules that do the actual work of
        disambiguation. By default, these files are isntalled into
        /usr/local/lib/site_perl/PERL_VERSION (where PERL_VERSION is the
        version of Perl you are using), or a similar directory. See the
        INSTALL file for more details.

    /bin
        This directory contains a script, wordtoset.pl, that lets you run
        the WSD software without writing your own Perl script.

    /doc
        This directory contains pod files for README, CHANGES, and INSTALL.
        These are what should be changed, the files found in the top level
        directory should be considered read-only.

    /t  This directory contains test scripts. These scripts are run when you
        run 'make test'.

SEE ALSO
    <http://senserelate.sourceforge.net/>

AUTHOR
     Ted Pedersen, University of Minnesota, Duluth
     tpederse at d.umn.edu

     Jason Michelizzi

    This document last modified by : $Id: README.pod,v 1.4 2008/03/22
    01:58:12 tpederse Exp $

COPYRIGHT AND LICENSE
    Copyright (c) 2004-2008, Ted Pedersen and Jason Michelizzi

    Permission is granted to copy, distribute and/or modify this document
    under the terms of the GNU Free Documentation License, Version 1.2 or
    any later version published by the Free Software Foundation; with no
    Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.

    Note: a copy of the GNU Free Documentation License is available on the
    web at <http://www.gnu.org/copyleft/fdl.html> and is included in this
    distribution as FDL.txt.



CHANGES
NAME
    CHANGES.pod

SYNOPSIS
    Revision history for Perl module WordNet::SenseRelate::WordToSet

DESCRIPTION
    0.03
    Released March 21, 2008 (all changes by TDP)

    *   added some general comments to README.pod

    *   modified wordtoset.pl to show the glosses of the word senses, in
        addition to the wps form

    *   expanded scope of test cases

    *   updated snyposis of WordToSet.pm with a cut, paste, and run example

    *   moved /util to /bin created /doc for .pod

    0.02
    Released June 12, 2005 (All changes by JM)

    *   fixed documentation to make it clear which options are required

    *   added PREREQ statement to Makefile.PL

    0.01
    Released Thursday May 12, 2005 (All changes by JM)

    *   original version; created by h2xs 1.23 with options -X
        WordNet::SenseRelate::WordToSet -b 5.6.0

AUTHORS
     Ted Pedersen, University of Minnesota, Duluth
     tpederse at d.umn.edu

    This document last modified by : $Id: CHANGES.pod,v 1.2 2008/03/22
    01:09:12 tpederse Exp $

SEE ALSO
COPYRIGHT AND LICENSE
    Copyright (c) 2004-2008 Ted Pedersen

    Permission is granted to copy, distribute and/or modify this document
    under the terms of the GNU Free Documentation License, Version 1.2 or
    any later version published by the Free Software Foundation; with no
    Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.

    Note: a copy of the GNU Free Documentation License is available on the
    web at <http://www.gnu.org/copyleft/fdl.html> and is included in this
    distribution as FDL.txt.



Makefile.PL
# $Id: Makefile.PL,v 1.2 2005/06/12 02:58:35 jmichelizzi Exp $
use strict;
use warnings;

use 5.006;
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'WordNet::SenseRelate::WordToSet',
    VERSION_FROM      => 'lib/WordNet/SenseRelate/WordToSet.pm', # finds $VERSION
    PREREQ_PM         => {'WordNet::QueryData' => 1.37,
		          'WordNet::Similarity' => 0.08},
    EXE_FILES         => ['utils/wordtoset.pl'],
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/WordNet/SenseRelate/WordToSet.pm', # retrieve abstract from module
       AUTHOR         => 'Jason Michelizzi <jmichelizzi at users.sourceforge.net>') : ()),
);


utils/wordtoset.pl
#!/usr/local/bin/perl

# $Id: wordtoset.pl,v 1.6 2005/06/12 03:01:41 jmichelizzi Exp $

my $Version = '0.02';

use strict;
use warnings;

use WordNet::SenseRelate::WordToSet;
use WordNet::QueryData;
use Getopt::Long;

my $measure = 'WordNet::Similarity::lesk';
my $trace = 0;
my $config;
my $version;
my $help;

my $res = GetOptions ("type|measure=s" => \$measure,
		      "trace=i" => \$trace,
		      "config=s" => \$config,
		      "version" => \$version,
		      "help" => \$help,
		      );

unless ($res) {
    showUsage();
    exit 1;
}

if ($version) {
    print "wordtoset.pl version ", $Version, "\n";
    print "Copyright (C) 2005, Jason Michelizzi and Ted Pedersen\n";
    print "This is free software, and you are welcome to redistribute it\n";
    print "under certain conditions.  This software comes with ABSOLUTELY\n";
    print "NO WARRANTY.  See the file COPYING or run 'perldoc perlgpl' for\n";
    print "more information.\n";
    exit;
}

if ($help) {
    showUsage("Long");
    exit;
}

my $target = shift;
my @context = @ARGV;
unless (defined $target and scalar @context) {
    unless (defined $target) {
	print STDERR "Error: no words were specified on command line!\n";
    }
    else {
	print STDERR "Error: no context words specified on command line!\n";
    }
    showUsage();
    exit 1;
}

my $qd = WordNet::QueryData->new;

# set up the options
my %options = (wordnet => $qd,
               measure => $measure,
	      );
$options{trace} = $trace if $trace;
$options{config} = $config if $config;

my $wts = WordNet::SenseRelate::WordToSet->new (%options);

my $result = $wts->disambiguate (target => $target,
				      context => [@context]);

for my $key (sort {$result->{$b} <=> $result->{$a}} keys %$result) {
  print $key, ' ', $result->{$key}, "\n";
}

if ($trace) {
    my $tstr = $wts->getTrace ();
    print "Traces:\n", $tstr, "\n";
}

sub showUsage
{
    my $long = shift;
    print "Usage: wordtoset.pl targetword contextword1 [contextword2 ...]\n";
    print "       [--type MEASURE] [--config FILE] [--trace LEVEL]\n";
    print "       [--help] [--version]\n";

    if ($long) {
	print "  --type MEASURE   WordNet::Similarity semantic relatedness measure\n";
	print "  --config FILE    Configuration file for relatedness measure\n";
	print "  --trace LEVEL    Turn tracing on/off\n";
	print "  --help           Show this help message\n";
	print "  --version        Show version information\n";
        print "You may also run 'perldoc' on this script for more details\n";
    }
}

__END__

=head1 NAME

wordtoset.pl - disambiguate a single word by comparing it to a set of words

=head1 SYNOPSIS

wordtoset.pl target context1 [context2 ...] [--type MEASURE] [--trace INT]
[--config FILE] | --help | --version

=head1 DESCRIPTION

wordtoset.pl performs word sense disambiguation on a single target word
by comparing the target word to a set of context words.  It serves as
a command line interface to the WordNet::SenseRelate::WordToSet Perl
module.

=head1 PARAMETERS & OPTIONS

=over

=item target

The target word to disambiguation.

=item context1 [context2...]

A list of words to use as context for disambiguating the target word.  No
order is presumed among the context words or among the context words and
the target word.

=item --type B<MEASURE>

The name of a WordNet::Similarity measure.  The default is
WordNet::Similarity::lesk.

=item --trace B<INTEGER>

Turn tracing on/off.  A value of zero turns tracing off, and a non-zero
value turns tracing on.  By default, tracing is off.  The trace levels
are:

  1 show non-zero scores from the semantic relatedness measure

  2 show zero & undefined scores from the relatedness measure (no
    effect unless combined with level 1)

  4 show traces from the semantic relatedness measure

The trace levels can be combined by adding together different levels.
For example, to show non-zero scores and the traces from the relatedness
measure, use level 5.

=item --config B<FILE>

The name of a configuration file for the specified semantic relatedness
measure.  See the documentation for the specific WordNet::Similarity
measure you are using to learn more about the format of a config file.

=item --version

Show version information.

=item --help

Show detailed help message.

=back

=head1 SEE ALSO

WordNet::SenseRelate::WordToSet(3)

L<http://senserelate.sourceforge.net/>

There are several mailing lists for SenseRelate:

L<http://lists.sourceforge.net/lists/listinfo/senserelate-users/>

L<http://lists.sourceforge.net/lists/listinfo/senserelate-news/>

L<http://lists.sourceforge.net/lists/listinfo/senserelate-developers/>

=head1 AUTHORS

Jason Michelizzi, E<lt>jmichelizzi at users.sourceforge.netE<gt>

Ted Pedersen, E<lt>tpederse at d.umn.eduE<gt>

=head1 BUGS

None known.

=head1 COPYRIGHT

Copyright (C) 2005 Jason Michelizzi and Ted Pedersen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


lib/WordNet/SenseRelate/WordToSet.pm
package WordNet::SenseRelate::WordToSet;

=head1 NAME

WordNet::SenseRelate::WordToSet - find sense of a target word most related to given set of words

=head1 SYNOPSIS

  use WordNet::SenseRelate::WordToSet;
  use WordNet::QueryData;
  my $qd = WordNet::QueryData->new;

  my %options = (measure => 'WordNet::Similarity::jcn',
                 wordnet => $qd);

  my $mod = WordNet::SenseRelate::WordToSet->new (%options);

  my $res = $mod->disambiguate (target => 'bank',
                              context => [qw/money cash dollar/]);

  # all senses for target and their scores are returned
  # we will just print the sense most related to the set

  $best_score = -100;
  foreach my $key (keys %$res) {
    next unless defined $res->{$key};
    if ($res->{$key} > $best_score) {
        $best_score = $res->{$key};
        $best = $key;
    }
  }

  # let's call WordNet::QueryData to get the gloss of the most
  # related sense of the target to the set 

  print "$best : ", join(", ", $qd->querySense($best, "glos")), "\n";

  my $res = $mod->disambiguate (target => 'bank',
                              context => [qw/river shore slope water/]);

  # all senses for target and their scores are returned
  # we will just print the sense most related to the set

  $best_score = -100;
  foreach my $key (keys %$res) {
    next unless defined $res->{$key};
    if ($res->{$key} > $best_score) {
        $best_score = $res->{$key};
        $best = $key;
    }
  }

  # let's call WordNet::QueryData to get the gloss of the most
  # related sense of the target to the set 

  print "$best : ", join(", ", $qd->querySense($best, "glos")), "\n";
  
=head1 DESCRIPTION

WordNet::SenseRelate::WordToSet finds the sense of a given target word 
that is most related to the words in a given set. 

=head2 Methods

The methods below will die() on serious errors.  Wrap calls to these
methods in an eval BLOCK to catch the exceptions.  See
'perldoc -f eval' for more information.

=over

=cut

use 5.006;
use strict;
use warnings;
use Carp;

our @ISA = ();
our $VERSION = '0.03';

my %wordnet;
my %simMeasure;
my %trace;
my %wnformat;
my %threshold;

# constants used to specify trace levels
#use constant TR_CONTEXT    =>  1;  # show the context window
#use constant TR_BESTSCORE  =>  2;  # show the best score
#use constant TR_ALLSCORES  =>  4;  # show all non-zero scores

# the previous three levels don't make a lot of sense for WordToSet
# * The context should be obvious
# * All the scores are returned from disambiguate() 
use constant TR_PAIRWISE   =>  1;  # show all the non-zero similarity scores
use constant TR_ZERO       =>  2;
use constant TR_MEASURE    =>  4;  # show similarity measure traces

=item B<new>Z<>

Z<>The constructor for this class.

Parameters:

  wordnet   => REFERENCE : WordNet::QueryData object (required)
  measure   => STRING    : name of a WordNet::Similarity measure (required)
  config    => FILENAME  : path to a config file for above measure
  trace     => INTEGER   : generate traces (default : 0)
  threshold => NUMBER    : similarity scores less than this are ignored

Returns:

   A reference to the constructed object or undef on error.

The trace levels are:

 1 show non-zero scores from the semantic relatedness measure

 2 show zero & undefined scores from the relatedness measure
   (no effect unless combined with level 1)

 4 show traces from the semantic relatedness measure

Note: the trace levels can be added together to achieve a combined effect.
For example, to show the non-zero scores, the zero scores, and the
traces from the measure, use level 7.

=cut

sub new
{
    my $class = shift;
    my %args = @_;
    $class = ref $class || $class;

    my $qd;
    my $measure;
    my $measure_config;
    my $threshold = 0;
    my $trace;
    my $wnformat = 0;

    while (my ($key, $val) = each %args) {
	if ($key eq 'wordnet') {
	    $qd = $val;
	}
	elsif ($key eq 'measure') {
	    $measure = $val;
	}
	elsif ($key eq 'config') {
	    $measure_config = $val;
	}
	elsif ($key eq 'threshold') {
	    $threshold = $val;
	}
	elsif ($key eq 'trace') {
	    $trace = $val;
	}
	elsif ($key eq 'wnformat') {
	    $wnformat = $val;
	}
	else {
	    croak "Unknown parameter type '$key'";
	}
    }

    unless (ref $qd) {
	croak "No WordNet::QueryData object supplied";
    }

    unless ($measure) {
	croak "No relatedness measure supplied";
    }

    my $self = bless [], $class;

    # initialize tracing
    if (defined $trace) {
	$trace{$self} = {level => $trace, string => ''};
	if (($trace & TR_ZERO) and !($trace & TR_PAIRWISE)) {
	    warn ("Warning: trace level ", TR_ZERO,
		  " has no effect unless combined with level ", TR_PAIRWISE,
		   ".\n");
	}
    }
    else {
	$trace{$self} = {level => 0, string => ''};
    }

    # setup relatedness measure
    my $file = $measure;
    $file =~ s/::/\//g;
    require "${file}.pm";

    if (defined $measure_config) {
	$simMeasure{$self} = $measure->new ($qd, $measure_config);
    }
    else {
	$simMeasure{$self} = $measure->new ($qd);
    }

    # check for errors
    my ($errCode, $errStr) = $simMeasure{$self}->getError;
    if ($errCode) {
        croak $errStr;
    }

    # turn on traces in the relatedness measure if required
    if ($trace{$self}->{level} & TR_MEASURE) {
        $simMeasure{$self}->{trace} = 1;
    }
    else {
        $simMeasure{$self}->{trace} = 0;
    }
    
    $wordnet{$self} = $qd;

    # store threshold value
    $threshold{$self} = $threshold;

    $wnformat{$self} = $wnformat;

    return $self;
}

sub DESTROY
{
    my $self = shift;
    
    delete $wordnet{$self};
    delete $simMeasure{$self};
    delete $threshold{$self};
    delete $trace{$self};
    delete $wnformat{$self};

    1;
}


=item disambiguate

Disambiguates the target word

Parameters:

  target  => STRING    : the target word to disambiguate (required)
  context => REFERENCE : a reference to an array of context words

Returns:

  A hash reference.  The keys of the hash will be the senses of the
  target word, and the values will be the score for each sense.

=cut

sub disambiguate
{
    my $self = shift;
    my %options = @_;

    # local vars
    my @context;
    my $target;

    while (my ($key, $val) = each %options) {
	if ($key eq 'target') {
	    $target = $val;
	}
	elsif ($key eq 'context') {
	    if ('ARRAY' eq ref $val) {
		@context = @$val;
	    }
	    else {
		carp "Value for option 'context' is not an array reference";
		return undef;
	    }
	}
	elsif ($key eq 'threshold') {
	    $threshold{$self} = $val;
	}
	else {
	    croak "Unknown option '$key'";
	}
    }

    my $tagged = 0;

    # quick sanity check to ensure that all words are in WordNet
    my $qd = $wordnet{$self};
    if ($wnformat{$self}) {
	foreach my $word ($target, @context) {
	    my @t = $qd->querySense ($word);
	    unless (scalar @t) {
		warn "'$word' is not found in WordNet\n";
		return undef;
	    }
	}
    }
    else {
	foreach my $word ($target, @context) {
	    my @t = $qd->validForms ($target);
	    unless (scalar @t) {
		warn "'$word' is not found in WordNet\n";
		return undef;
	    }
	}
    }
    
    my $result;
    $result = $self->doNormal ($target, @context);

    return $result;
}

sub doNormal
{
    my $self = shift;
    my $target = shift;
    my @context = @_;
    my $measure = $simMeasure{$self};
    my $threshold = $threshold{$self};

    my $tracelevel = $trace{$self}->{level};
    my @traces;

    # get senses for the target and context words
    my @targetsenses = $self->_getSenses ($target);
    my @contextsenses;
    for my $i (0..$#context) {
	$contextsenses[$i] = [$self->_getSenses ($context[$i])];
    }


    # now disambiguate the target

    my @sums;
    for my $targetsense (0..$#targetsenses) {
	$sums[$targetsense] = 0;

	for my $i (0..$#contextsenses) {
	    next if 0 == scalar $contextsenses[$i];
	    my @tempScores;

	    for my $k (0..$#{$contextsenses[$i]}) {
		unless (defined $contextsenses[$i][$k]) {
		    warn "\$contextsenses[$i][$k] is undef";
		}

		$tempScores[$k] = 
		    $measure->getRelatedness ($targetsenses[$targetsense],
					      $contextsenses[$i][$k]);
	    }

	    my $max = -1;
	    my $maxidx = -1;
	    for my $n (0..$#tempScores) {
		if ($tracelevel & TR_PAIRWISE) {
		    if (($tempScores[$n] && $tempScores[$n] > 0)
			|| ($tracelevel & TR_ZERO)) {
			unless (defined $contextsenses[$i][$n]) {
			    warn "\$contextsenses[$i][$n] is undef";
			}
			my $s = "    "
                            . $targetsenses[$targetsense] . ' '
                            . $contextsenses[$i][$n] . ' '
                            . (defined $tempScores[$n]
                               ? $tempScores[$n]
                               : 'undef');

			push @{$traces[$targetsense]}, $s;
		    }
		}

		if ($tracelevel & TR_MEASURE) {
		     if (($tempScores[$n] && $tempScores[$n] > 0)
			|| ($tracelevel & TR_ZERO)) {
			 push @{$traces[$targetsense]}, $measure->getTraceString;
		     }
		}

		$measure->getError; # clear errors from relatedness object

		if (defined $tempScores[$n] && ($tempScores[$n] > $max)) {
		    $max = $tempScores[$n];
		    $maxidx = $n;
		}

	    }
	    
	    $sums[$targetsense] += $max if $max > $threshold;
	}
    }

    my $max = -1;
    my $maxidx = -1;
    foreach my $p (0..$#sums) {
	if ($sums[$p] > $max) {
	    $maxidx = $p;
	    $max = $sums[$p];
	}

#	if ($tracelevel & TR_ALLSCORES
#	    && (($sums[$p] > 0) or ($tracelevel & TR_ZERO))) {
#	    $trace{$self}->{string} .= "   $targetsenses[$p]: $sums[$p]\n";
#	}

	if (($tracelevel & TR_MEASURE or $tracelevel & TR_PAIRWISE)
	    && defined $traces[$p]) {
	    for my $str (@{$traces[$p]}) {
		$trace{$self}->{string} .= $str . "\n";
	    }
	}
    }

	

    my %rhash;
    my $best_sense = '';
    my $best_score = -1;
    foreach my $i (0..$#sums) {
	if ($sums[$i] > $best_score) {
	    $best_sense = $targetsenses[$i];
	    $best_score = $sums[$i];
	}

	$rhash{$targetsenses[$i]} = $sums[$i] if $sums[$i] > $threshold;
    }

#    if ($tracelevel & TR_BESTSCORE) {
#	if ($best_score >= 0) {
#	    $trace{$self}->{string} .= " Winning sense: $best_sense\n";
#	    $trace{$self}->{string} .= " Winning score: $best_score\n";
#	}
#	else {
#	    $trace{$self}->{string} .= " Winning sense: (none)\n";
#	    $trace{$self}->{string} .= " Winning score: (none)\n";
#	}
#    }

    return \%rhash;

#    if ($maxidx >= 0) {
#	return $targetsenses[$maxidx];
#    }

#    return $target;
}

=item B<getTrace>

Gets the current trace string and resets it to "".

Parameters:

None

Returns:

The current trace string (before resetting it).  If the returned string
is not empty, it will end with a newline.

Example:

  my $str = $wsd->getTrace ();
  print $str;

=cut

sub getTrace
{
    my $self = shift;

    return '' unless $trace{$self};

    my $s = $trace{$self}->{string};
    $trace{$self}->{string} = '';
    return $s;
}

sub _getSenses
{
    my $self = shift;
    my $word = shift;
    my $qd = $wordnet{$self};
    my @senses;

    # first get all forms for each POS
    if ($word =~ /\#o/) {
	@senses = undef;
    }
    else {
	my @forms;
	unless ($wnformat{$self}) {
	    @forms = $qd->validForms ($word);
	}
	else {
	    @forms = $word;
	}
	
	if (scalar @forms == 0) {
	    @senses = ();
	}
	else {
	    # now get all the senses for each form
	    foreach my $form (@forms) {
		my @temps = $qd->querySense ($form);
		push @senses, @temps;
	    }
	}
    }

    return @senses;
}


1;

__END__

=back

=head1 SEE ALSO

L<http://senserelate.sourceforge.net/>.

 Ted Pedersen, Satanjeev Banerjee, and Siddharth Patwardhan (2005)
 Maximizing Semantic Relatedness to Perform Word Sense Disambiguation,
 University of Minnesota Supercomputing Institute Research Report UMSI
 2005/25, March.
 L<http://www.msi.umn.edu/general/Reports/rptfiles/2005-25.pdf>

=head1 AUTHORS

Ted Pedersen, University of Minnesota, Duluth
tpederse at d.umn.edu

Jason Michelizzi

Last modified by :
$Id: WordToSet.pm,v 1.8 2008/03/22 01:14:04 tpederse Exp $

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2008 by Jason Michelizzi and Ted Pedersen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  
USA

=cut
